### PR TITLE
feat(import): Import YouTube playlist from Nightbot

### PR DESCRIPTION
**Pull request by redbeardymcgee**
This commit implements the first pass at importing the playlist managed by Nightbot. In the code, I have left a few comments to make note of some undercooked solutions.

I tried doing this with simple `React.useState()` but you may find that this approach is too naive. For now, I do it this way to make it easier to output the playlist resource into a text field. It's something like a `console.log(playlist)` with extra steps.

Setting aside code style, my next step is to figure out how to convert the tracks from Nightbot into tracks for SogeBot. Do you have some pointers to where I could start gluing that together? 

In [playlist.tsx](https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L263) I see a handler for adding songs to the playlist, but I'm not sure if that's the right inspiration I should find (copy and paste lmao).

Also, existing SogeBot users may discover this feature and wish to import their old Nightbot playlist. There is some potential for such an import to clobber an existing playlist, I think by modifying `createdAt` or `updatedAt` values or duplicating items. I'm not sure how to mitigate that yet.

### CODE REVIEW

**Comment by sogehige** [comment]
> In [playlist.tsx](https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L263) I see a handler for adding songs to the playlist, but I'm not sure if that's the right inspiration I should find (copy and paste lmao).

Yes, you should use https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L271C8-L273C10

e.g.
```typescript
const async import = () => {
  const videoIds = ['bvNZeh6f8vE']
  for (const videoId of videoIds) {
    await new Promise((resolve, reject) => getSocket('/systems/songs').emit('import.video', {
      playlist: videoId, forcedTag: 'nightbot-import'
    }, (err) => {
      if (err) {
        reject(err)
      } else {
        resolve()
      }
    }))
  };
}
```
        
it may take a while as we are fetching video data, so thats why it is one by one, you should handle reject as well as some videos might not be available anymore

**Comment by redbeardymcgee** [comment]
Due to the amount of overall churn in the diffs because of my note-taking in comments, I find this code easier to look at through this [link to the file](https://github.com/sogebot/sogebot.dev/blob/888b35526a882b613c0086eb15a956f80c38828f/ui.admin-mui/src/components/Settings/Import/nightbot.tsx#L69).

I will be removing all comments as the implementation becomes more stable. Until then, they include open questions that may answer themselves, again a result of the implementation details stabilizing. If you do have some input on them, feel free to share here.

Thank you!

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 1]
```diff
@@ -65,29 +66,156 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+
+    return pl;
+  };
+
+  const importPlaylist = async () => {
+    // TODO: `fetchPlaylist` should probably just return the simplified array of IDs
+    const videos: PlaylistItemTrack[] = await fetchPlaylist();
+
+    videos.filter((track) => {
+      // TODO: Nightbot supports SoundCloud also
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: no, we are not supporting soundcloud at all

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 2]
```diff
@@ -65,29 +66,156 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+      await new Promise((resolve, reject) => {
+        getSocket('/systems/songs').emit(
+          'import.video',
+          {
+            playlist:  video.providerId,
+            forcedTag: 'nightbot-import',
+          },
+          (err) => {
+            if (err) {
+              // FIXME: This could potentially generate
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: you can just count number of failures and if >0 then just enqueueSnackbar something like `x video(s) failed to import` or something like that. wdyt?

**Comment by redbeardymcgee** [comment]
This is effectively a final draft of the implementation. What remains now are stylistic or organization issues, or bugs.

### Potential Bugs
 - Type checker returns "Cannot find name 'T'. [2304]" for `fetchPlaylist`
 - That's all, obviously. :)
 - Should I even attempt retries or just report failure, log it, and bail out?

### Style
 - `fetchPlaylist()` is recursive. Is it unclear in this style?
 - 3 instances of mutation: `maxRetries`, `acc.push()`, `failCount`
 - The types may be unnecessary, or should be moved to another file maybe?
 - I should at least log somewhere instead of a snackbar or console, right?

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 3]
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+  type PlaylistPage = {
+    status: number;
+    _sort: { date: 'asc' | 'desc' };
+    _limit: number;
+    _offset: number;
+    _total: number;
+    playlist: PlaylistItem[];
+  };
+
+  const fetchPlaylist = async (acc: PlaylistItemTrack[], offset: number): Promise<T> => {
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
  const fetchPlaylist = async (acc: PlaylistItemTrack[] = [], offset = 0): Promise<T> => {
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 4]
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+        console.error(`Error === ${error.response.status}`);
+      } else {
+        enqueueSnackbar('Remote server error.', { variant: 'error' });
+        console.error(`Error !== 429 or 5XX: ${error.response.status}`);
+      }
+    }
+  };
+
+  const importPlaylist = async () => {
+    const videos: PlaylistItemTrack[] = await fetchPlaylist([], 0);
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
    const videos: PlaylistItemTrack[] = await fetchPlaylist();
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx (hunk 3 above)
Comment (sogehige) [reply ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: seems little bit nicer for me, but your version is ok as well, depends on you

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 5]
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+          label={translate('integrations.lastfm.settings.username')}
+          InputProps={{
+            endAdornment: <InputAdornment position='end'> { user !== 'Not Authorized'
+              ? <Button color='error' variant='contained' onClick={revoke}>Revoke</Button>
+              : <Button color='success' variant='contained' onClick={authorize}>Authorize</Button>
+            }
+            </InputAdornment>,
+          }}
+        />
+        <Stack direction='row-reverse'>
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
        <Stack direction='row'>
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 6]
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+          <Button color='primary' variant='contained' disabled={user === 'Not Authorized'} onClick={importPlaylist}>Import</Button>
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
          <Button color='primary' variant='contained' disabled={user === 'Not Authorized'} onClick={importPlaylist}>Import playlist</Button>
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 7]
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+      const data: PlaylistPage = response.data;
+      const tracks = data.playlist.map((e) => e.track);
+
+      if (acc.push(...tracks) >= data._total ?? 0) {
+        return acc;
+      }
+
+      return await fetchPlaylist(acc, offset + 100);
+    } catch (error: any) {
+      if (error.response.status === 429 && maxRetries !== 0) {
```
Comment (redbeardymcgee) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: I think the conditional branches in here should also return the accumulator. Currently i think both else branches just fall out, returning `undefined` as a default.

**Comment by redbeardymcgee** [comment]
I think this is in a workable state now. The feature works fine, but of course I'm worried that I messed something up. I will remain available in case someone finds a way to break my implementation, so I can try to fix it better :)

**Comment by redbeardymcgee** [comment]
I was testing this out and it's no longer catching the 429. Have to rework the error handling.

**Comment by redbeardymcgee** [comment]
Copying the commit message body here for visibility:

> Previously, the retry logic was naive in a way that made it do nothing.
> While it technically did retry a failed request, it only retried a
> single time, and the result of the retry was discarded.
> 
> This version actually captures error responses and retries each failed
> request up to 3 times, delayed by 1 minute each. This will block the
> importer from receiving the playlist tracks from Nightbot for
> potentially a long time, but will also gracefully handle 429 errors.
> 
> Using a generator function may allow for processing a page of tracks
> while other requests are awaiting a response. It is unlikely such an
> optimization will be necessary, since this feature is expected to be
> used only 1 time per user. Additionally, the number of users who will
> care to import their playlist from Nightbot is probably very small.

Now there is the matter of http 410 errors from YouTube when a video has been delisted. The error happens during the import step on the backend, which I have been hesitant to touch. Previously, I thought it was just being discarded by SogeBot, which is probably the best thing to do aside from logging the error. It doesn't seem correct for me to make a request to YouTube to check that the video is still available, since SogeBot will then need to make a second request for the same video ID when importing. Does that make sense?

**Comment by redbeardymcgee** [comment]
For now I've chosen to swallow the errors returned by the video import, since there really isn't some useful thing to do when it fails. Almost certainly it's a result of a delisted video.

If you have no objections, I think this can be considered complete. Please let me know if there's anything else I can add or change.

**Pull request merged**